### PR TITLE
Pass RobotModel to kinematics plugins

### PR DESCRIPTION
**Pull request by jonbinney**
Instead of just passing in the URDF as a string. This is based on the discussion in https://github.com/ros-planning/moveit_ros/issues/731 and is just a rebasing of @davetcoleman's PRs https://github.com/ros-planning/moveit_core/pull/216 and https://github.com/ros-planning/moveit_ros/pull/541

There are two main motivations for this:
1. Parsing the URDF string multiple times is wasteful
2. The RobotModel actually has more information than the URDF. For example, it can have modified max/min joint positions that were specified as ROS parameters.

This doesn't actually fix (2) yet. This PR is the first step - it modifies the KinematicsBase API to pass the RobotModel to the kinematics plugins. A future PR will need to then use the values from the RobotModel instead of the URDF in each separate kinematics plugin.

I've rebased the original PRs and fixed the compile problems, but now I'm not sure how to test this. Ideally, there would be unit/integration tests that would load various plugins (in this case the KDL and SRV kinematics plugins) and make sure that they return valid answers for various kinematic and inverse kinematic queries for various robot arms. I was hoping that https://github.com/ros-planning/moveit_kinematics_tests/issues/4 contained such tests, but it looks like it isn't (yet?) used for that purpose.

So my question is - how do people test things like this to make sure it works for various robots?

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Dave's PRs didn't touch it either (because I don't think it existed at the time), but I think the ('recently' added) [LMA based plugin](https://github.com/ros-planning/moveit/tree/kinetic-devel/moveit_ros/planning/lma_kinematics_plugin) would also need to be updated?

**Comment by davetcoleman** [comment]
This is great!

@jrgnicho can hopefully expand upon moveit_kinematic_tests

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 1]
```diff
@@ -388,6 +391,8 @@ class KinematicsBase
    * @param tip_frame The tip of the chain
    * @param search_discretization The discretization of the search when the solver steps through the redundancy
    * @return True if initialization was successful, false otherwise
+   *
+   * Note: this is intended to be removed in future ROS versions in favor of robot_model being passed in
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: lets add MOVEIT_DEPRECATED then

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 2]
```diff
@@ -416,17 +423,39 @@ class KinematicsBase
...
     return false;
   }
 
   /**
+   * @brief  Initialization function for the kinematics, for use with kinematic chain IK solvers
+   * @param robot_model - allow the URDF to be loaded much quicker by passing in a pre-parsed model of the robot
+   * @param group_name The group for which this solver is being configured
+   * @param base_frame The base frame in which all input poses are expected.
+   * This may (or may not) be the root frame of the chain that the solver operates on
+   * @param tip_frame The tip of the chain
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: tip_frames plural (i know im now reviewing my own code from a couple years ago!)

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 3]
```diff
@@ -416,17 +423,39 @@ class KinematicsBase
...
+   * @param search_discretization The discretization of the search when the solver steps through the redundancy
+   * @return True if initialization was successful, false otherwise
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: True lowercase

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 4]
```diff
@@ -416,17 +423,39 @@ class KinematicsBase
...
+   */
+  virtual bool initialize(const moveit::core::RobotModel* robot_model,
```
Comment (rhaschke) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: I think it is dangerous to pass in a pure pointer here. Previously the kinematics plugin created and owned the RobotModel. If we pass in a pointer, it should be a shared_ptr to ensure existence of the RobotModel for the whole lifetime of the plugin.

File: moveit_ros/planning/srv_kinematics_plugin/src/srv_kinematics_plugin.cpp [hunk 5]
```diff
@@ -71,19 +72,15 @@ bool SrvKinematicsPlugin::initialize(const std::string &robot_description,
...
-  const boost::shared_ptr<srdf::Model> &srdf = rdf_loader.getSRDF();
-  const boost::shared_ptr<urdf::ModelInterface>& urdf_model = rdf_loader.getURDF();
 
-  if (!urdf_model || !srdf)
+  // Check robot_state is initialized
+  if (!robot_state_)
   {
-    ROS_ERROR_NAMED("srv","URDF and SRDF must be loaded for SRV kinematics solver to work."); // TODO: is this true?
-    return false;
+    robot_state_.reset(new robot_state::RobotState(robot_model->getConstPtr()));
```
Comment (rhaschke) [inline moveit_ros/planning/srv_kinematics_plugin/src/srv_kinematics_plugin.cpp]: Ah. You actually store a shared_ptr only. Wouldn't it make sense to directly pass in the shared_ptr?
This would clarify the semantics from an inspection of the headers and it will avoid race conditions: What if robot_model was destroyed just before the `getConstPtr()` call?

**Comment by jrgnicho** [comment]
The kinematics_base_test repo was created to ensure that changes to the ``kinematics::KinematicsBase`` interface did't break existing implementation ( see my comments [here](https://github.com/ros-planning/moveit_kinematics_tests/issues/4) ).  It can be used to test the changes in this PR.

File: moveit_ros/planning/srv_kinematics_plugin/src/srv_kinematics_plugin.cpp (hunk 5 above)
Comment (davetcoleman) [inline moveit_ros/planning/srv_kinematics_plugin/src/srv_kinematics_plugin.cpp]: It is because the shared_ptr isn't available during construction of the class - that's why I had to use ``: public boost::enable_shared_from_this<RobotModel>``

This code was from two years ago, so perhaps I missed something back then and there's a way to do this better, but I'm pretty sure there wasn't

Comment (jonbinney) [inline moveit_ros/planning/srv_kinematics_plugin/src/srv_kinematics_plugin.cpp]: Does this create a `shared_ptr` loop? `RobotModel` contains `JointModelGroup`, which has the kinematics solver in the `group_kinematics_` member variable. Since this kinematics solver will now contain a `shared_ptr` to a `RobotState` which keeps a `shared_ptr` to the `RobotModel`, doesn't that create a loop?

Comment (davetcoleman) [inline moveit_ros/planning/srv_kinematics_plugin/src/srv_kinematics_plugin.cpp]: I don't believe so, I used this code for about a year without any signs of a loop

Comment (jonbinney) [inline moveit_ros/planning/srv_kinematics_plugin/src/srv_kinematics_plugin.cpp]: ok. I need to investigate a bit further into what is created/owned by who and where, and I'll update this PR with a summary in a couple days.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 1 above)
Comment (jonbinney) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: I tried adding MOVEIT_DEPRECATED to both this function and the other version of initialize() that uses the string urdf below. gcc spits out tons of warnings though, complaining about the below deprecated initialize() calling this deprecated initialize. This is on ubuntu 16.04 with `gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.2) `.

Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: ah right, ok

Comment (jonbinney) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: It seems like that shouldn't be the case though - if a deprecated function calls another deprecated function, shouldn't that be OK? Any C++ gurus that know a workaround?

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 2 above)
Comment (jonbinney) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: fixed

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 3 above)
Comment (jonbinney) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: fixed

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 1 above)
Comment (jonbinney) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: @baalexander suggested marking the below one as deprecated now, and then later removing it and marking this one as deprecated. That would take longer, but would have the advantage of not spewing warnings.

**Comment by jonbinney** [comment]
@gavanderhoorn I'll take a look at updating the LMA based plugin. That should definitely be included in this PR.

**Comment by jonbinney** [comment]
I've made some progress on this PR but still have some more work to do before it is ready to be reviewed again. The biggest problem remaining is that there is definitely a loop of shared pointers that keeps the RobotModel from ever being destroyed. The problem is:

- `RobotModel` contains `JointModelGroup`, 
- `JointModelGroup` has the kinematics solver in the `group_kinematics_` member variable
- The kinematics solver contains a shared_ptr to a `RobotState`
- `RobotState` contains a shared_ptr to the `RobotModel`

So the loop is `RobotModel->JointModelGroup->KDLKinematicsPlugin->RobotState->RobotModel`.

Any suggestions on how to fix this? It seems like one of these shared pointers needs to change to a weak pointer, but I'm not sure which.

**Comment by v4hn** [comment]
> Any suggestions on how to fix this? It seems like one of these shared pointers needs to change to a weak pointer, but I'm not sure which.

With the current design, I suppose `KDLKinematicsPlugin->RobotState` is the best candidate,
because it points upwards in MoveIt's hierarchy.

**Comment by rhaschke** [comment]
The reason @v4hn gives to motivate the `KDLKinematicsPlugin->RobotState` as a `weak_ptr` candidate is not clear to me. IMHO, we should look for weak couplings of instances. `KDLKinematicsPlugin -> RobotState` is the worst candidate: `KDLKinematicsPlugin` owns these `RobotState` pointers!

@jonbinney, there are many other cycles, e.g. `RobotModel->JointModelGroup->RobotModel`. In most cases, some "child class" stores a pointer back to its "parent class", here `JointModelGroup` stores a pointer to `RobotModel` in `parent_model_`. In this case this back pointer should be the weak one. Probably it could be even replaced by a proper reference: The pointer should never be invalid...

I think, the best way to break the loop in the above case, is to pass a new `RobotModelSharedPtr` to `RobotState`: As long as the `KDLKinematicsPlugin` is alive, its parent RobotModel will be so too. Hence, we could pass in the same pointer (`.get()`), but with a no-deleter.

**Comment by v4hn** [comment]
> The reason @v4hn gives to motivate the KDLKinematicsPlugin->RobotState as a weak_ptr candidate is not clear to me. IMHO, we should look for weak couplings of instances. KDLKinematicsPlugin -> RobotState is the worst candidate: KDLKinematicsPlugin owns these RobotState pointers!

ok, I didn't read through the dependencies in the recent past, so feel free to ignore my statement. :baby:

**Comment by davetcoleman** [comment]
ping @jonbinney any chance we could finish up this PR?

**Comment by jonbinney** [comment]
Ah yes, sorry! This fell off my radar. I'll start working on it again.

**Comment by jonbinney** [comment]
Having fun merging these changes with all the clang-format changes....

**Comment by jonbinney** [comment]
I haven't had any time to work on this for a while, and unfortunately I'm going to be very busy for the next few months as well. I'm going to close this for now. When/if I have time in the future I'll re-open it (or if anyone else has some time they're welcome to take it over). The changes in this PR are good, but time will need to be put in to make sure that they don't break existing behavior, since it affects a lot of code. Things that need to be resolved:
- merge/rebase to account for changes in kinetic-devel
- fix shared pointer cycles
- update some of the inverse kinematic plugins and create a guide to updating others
- lots of testing

**Comment by 130s** [comment]
I'd suggest to keep this open -- closed issue & PRs are harder to be found.

**Comment by v4hn** [comment]
I agree with @130s, there's no need to close the request.
With an open request, the topic pops up every now and then.
Without it, especially new contributors will never notice that work in this direction already exists.

I'll reopen this request. @jonbinney if you don't want to have the open request on your name around or don't want to keep the branch on your repository I can transition the commits to my fork. But I don't have the time to work on this either.

**Comment by jonbinney** [comment]
Makes sense - I'm fine with keeping it open.

**Comment by simonschmeisser** [comment]
@v4hn you mentioned on Sunday that this PR has been discussed at ROScon, could you summarize the results? I'll try to look into this soonish and would like to be up to date beforehand

I might just take the changes here as an inspiration and do a new PR without any rebasing etc. My git skills are weak. Would that offend anyone?

**Comment by jonbinney** [comment]
@simonschmeisser thanks for taking this on! Yes, feel free to start fresh - no need to preserve the history here.

One thought - it would be good to have a shim set of functions that keep the current interface, construct the robot model from the URDF, and then call the new functions. We can mark them as deprecated so people have some warning and their custom IK plugins don't just stop working.

**Comment by simonschmeisser** [comment]
Just had a discussion with @v4hn about this at IPA at WorldMoveItDay:

It seems the best way to cut the shared_ptr dependency loop is to change the `kinematics::KinematicsBasePtr solver_instance_` in `JointModelGroup::KinematicsSolver` JointModelGroup.h to a `std::weak_ptr`

I have a slight fear that I will not finish that within World MoveIt! Day (any time zone) but I will start a PR testing that implementation.

**Comment by rhaschke** [comment]
> It seems the best way to cut the shared_ptr dependency loop is to change the kinematics::KinematicsBasePtr solver_instance_ in JointModelGroup::KinematicsSolver JointModelGroup.h to a std::weak_ptr

Can you please elaborate on this conclusion? I tend to agree, but it would be nice to understand (and document) your motivation as well ;-)

**Comment by rhaschke** [comment]
Please also get rid of `JointModelGroup::solver_instance_const_`. There is no need for this. A `shared_ptr<T>` can be casted into a `shared_ptr<const T>`.

**Comment by simonschmeisser** [comment]
Sure, sorry for not giving more details.

Your ( @rhaschke ) suggestion was to create a special std::shared_ptr<RobotModel, doNotDeleteMe>(robotModel.get()) and pass that to RobotState instances in the IK plugins. While this (probably) works, there would no direct way to stop people from using the "normal" RobotModel shared ptr and create that loop. We had some ideas for hacks/factory patterns etc but ...

On the other hand all kinematic plugin instances are stored in KinematicsPluginLoader/RobotModelLoader as shared_ptr already. So we can use weak_ptrs elsewhere, especially in JointModelGroup

Making the RobotModel pointer of RobotState weak would also have been possible but too invasive


re const, yes, thought that as well

**Comment by rhaschke** [comment]
> Your suggestion was to create a special `std::shared_ptr<RobotModel, doNotDeleteMe>(robotModel.get())` and pass that to `RobotState` instances in the IK plugins. ... There would be no direct way to stop people from using the "normal" `RobotModelPtr` and create that loop.

I don't understand what you mean by "normal" here. I looked into this issue again for last hour and found the following:

The RobotModelLoader (RML) owns its RobotModel (RM) as well as its KinematicsPluginLoader (KPL).
The JointModelGroup (JMG) stores the actual solver instance (KS), which is *also* cached in KPL.
```
RML => RM => JMG => KS
        \------<----/
```
Usually those caches are the proper choice for weak ptrs: As long as the pointer is still actively in use (outside the cache), the cached pointer can be reused. If all active instances are gone, the cache instance should be freed as well. However, in the present case, the cache only maintains unique instances - they are not shared. The only purpose of this logic is to not allocate KS twice in a row - once for validating that it is suitable for JMG [here](https://github.com/ros-planning/moveit/blob/a46f0fb654dee077bcf39c1a2c48f4dffe1068f4/moveit_ros/planning/robot_model_loader/src/robot_model_loader.cpp#L203) and once for actual use [here](https://github.com/ros-planning/moveit/blob/a46f0fb654dee077bcf39c1a2c48f4dffe1068f4/moveit_core/robot_model/src/joint_model_group.cpp#L604). The important fact is, that not the KPL _owns_ those instances, but the JMGs! **Hence, the JMGs should store the shared_ptr.** By clearing KPL's cache (and only the cache!) before RM in RML's destructor this can be handled. A better solution is to std::move the cached shared_ptr for return, but keeping the cache entry (which then has `use_count = 0`).

Regardless of this, the loop isn't broken, because the RM and thus JMG is not destroyed in first place, when the RML is unloaded. Thus, we need to ensure that the RML, in its destructor, actually triggers the destructor pipeline RM => JMG => KS. Hence, the only chance to break the loop is from KS _back to_ RM.
(We cannot break the loop at arbitrary position, but we have to at (all?) back pointers.)

As far as I see, the old code doesn't provide an interface to directly pass a RobotModelPtr to KS.
KDLKinematicsPlugin, creates its own, new instance [here](https://github.com/ros-planning/moveit/blob/a46f0fb654dee077bcf39c1a2c48f4dffe1068f4/moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp#L144), which we want to avoid with this PR. Thus, my suggestion is to break the loop at this new interface, not passing in a shared_ptr, but a reference. This states, that the fact that the pointer is weak. Passing in a weak_ptr, would not prevent KS authors to turn it into a proper shared_ptr and store it again. Instead, KS authors need to create their own "dummy" shared_ptr from this reference via `std::shared_ptr<RobotModel, noDeleter>(&model)`. That's exactly the same conclusion I found in https://github.com/ros-planning/moveit/pull/150#issuecomment-254240560.

@simonschmeisser Sorry for the late reply. But it took me a while to understand all this. I suggest, I propose a solution along this line? (I still have all files open...)

**Comment by rhaschke** [comment]
I committed a solution to a similar issue in `KinematicsPluginLoader` [here](https://github.com/ros-planning/moveit/pull/1104/commits/71821e43b0b252052090ba16f39b6f5a097ce7de) - https://github.com/ros-planning/moveit/pull/1104#issuecomment-433073694.
@simonschmeisser I didn't continued here, as I didn't get feedback from you. I guess you already rebased this PR to melodic-devel?

**Comment by simonschmeisser** [comment]
Please see my comments on #1153 and consider reopening it

**Comment by jonbinney** [comment]
https://github.com/ros-planning/moveit/pull/1166 and https://github.com/ros-planning/moveit/pull/1153 are more recent attempts at this, so I'll close this PR.
